### ISSUE TITLE
New Module Request : MSSQL Audit Module

### ISSUE THREAD

**Issue by maxverro** [elastic/beats#20001]
Hi,

This is a request to add a new MSSQL Audit Module possibly to Winlogbeat as it would make sense since MSSQL Audit is published directly to the Windows Event Log. If this is something that you see as a possibility for Winlogbeat I could summit a PR.

## Use case
To enable our SOC team staff to detect and respond in a constant manner to possible malicious MSSQL Server changes or malicious system command run with xp_cmdshell during unexpected times or from unexpected users we need to collect these logs. Once indexed in Elasticsearch the SOC team can use them to produce signals and take action directly from the SIEM app. 

## Suggestion
Every log produced by MSSQL Server is in the same format, output should be fairly consistent, here is a sample : 

```
Audit event: audit_schema_version:1
event_time:2020-07-16 15:01:56.1739520
sequence_number:1
action_id:EX  
succeeded:true
is_column_permission:false
session_id:1
server_principal_id:201
database_principal_id:1
target_server_principal_id:0
target_database_principal_id:0
object_id:-1008133434
user_defined_event_id:0
transaction_id:0
class_type:X 
permission_bitmask:00000000000000000000000000000020
sequence_group_id:9A54C953-8729-459A-A8FC-4KF2D8C2
session_server_principal_name:mssql-01\admin
server_principal_name:mssql-01\admin
server_principal_sid:0105000000000005151111006ffd01a97582952d715b5c39f4010000
database_principal_name:dbo
target_server_principal_name:
target_server_principal_sid:
target_database_principal_name:
server_instance_name:mssql-01
database_name:master
schema_name:sys
object_name:xp_cmdshell
statement:EXEC xp_cmdshell 'dir *.exe'
additional_information:
user_defined_information:
.
```

#### Mapping fields to ECS suggestion

Event Viewer Field Name | Converted Field Name (ECS) | Sample Value | Comment
-- | -- | -- | --
event_time | @timestamp | 2020-07-16 15:32:22.7140549 |  
sequence_number | event.sequence | 1 |  
action_id | event.action | execution | Value map can be obtain with `select * from sys.dm_audit_actions`
succeeded | event.outcome | success | success or failed. Winlogbeat already those that.
  | event.category | database  | database https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html
  | event.type | access  | access or change https://www.elastic.co/guide/en/ecs/current/ecs-allowed-values-event-category.html
is_column_permission | mssql.audit.is_column_permission | false |  
session_id | mssql.audit.session_id | 282 |  
server_principal_id | mssql.audit.server_principal_id | 523 |  
database_principal_id | mssql.audit.database_principal_id | 14 |  
target_server_principal_id | mssql.audit.target_server_principal_id | 0 |  
target_database_principal_id | mssql.audit.target_database_principal_id | 0 |  
object_id | mssql.audit.object_id | 0 |  
user_defined_event_id | mssql.audit.user_defined_event_id | 0 |  
transaction_id | mssql.audit.transaction_id | 6798281735 |  
class_type | mssql.audit.class_type | SR | Value map can be obtain with `select * from sys.dm_audit_class_type_map`
permission_bitmask | mssql.audit.permission_bitmask | 00000000000000000000000000000000 |  
sequence_group_id | mssql.audit.sequence_group_id | BDA21044-0D73-4E7C-908D-5C7B82504C3D |  
session_server_principal_name | mssql.audit.session_server_principal_name | hackme |  
server_principal_sid | mssql.audit.server_principal_sid | 0d271d7fbd1ea84fb9042d5fc4468420 |  
database_principal_name | mssql.audit.database_principal_name | hackme |  
target_server_principal_name | mssql.audit.target_server_principal_name |   |  
target_server_principal_sid | mssql.audit.target_server_principal_sid |   |  
target_database_principal_name | mssql.audit.target_database_principal_name |   |  
server_instance_name | mssql.audit.server_instance_name | mssql-01 |  
database_name | mssql.database.name | master |  
schema_name | mssql.database.schema_name | sys |  
object_name | mssql.database.object_name | xp_cmdshell |  
statement | mssql.audit.statement | EXEC xp_cmdshell 'dir *.exe' |  `mssql.statement` could be possible. It is used in the APM I think
additional_information | mssql.audit.additional_information | <tsql_stack><frame nest_level = '1' database_name = 'TEST_DB' schema_name = 'TEST_SCHEMA' object_name = 'test_object'/></tsql_stack> |  
user_defined_information | mssql.audit.user_defined_information | . |  

## Setting up environment 
For this to work, end users must configure **MSSQL Server Audit** with some **Audit Specification**. A sample configuration would look like this. 

```sql
-- Create and activate server audit policies 

USE master

CREATE SERVER AUDIT SIEM

TO APPLICATION_LOG

WITH

(QUEUE_DELAY = 1000, ON_FAILURE =

CONTINUE

)

ALTER SERVER AUDIT SIEM

WITH (STATE = ON)

 

-- Server: Audit server configuration changes

-- Windows Log: Application

-- Events: 15457 

CREATE SERVER AUDIT SPECIFICATION [Audit_Server_Configuration_Changes]

FOR SERVER AUDIT SIEM

ADD (AUDIT_CHANGE_GROUP),                     -- Audit changes to audit policies

ADD (SERVER_OPERATION_GROUP)                  -- Audit server level changes

WITH (STATE = ON)

 

-- DATABASE: Audit common agent job activity

-- Windows Log: Application

-- Events: 33205 

Use msdb

CREATE DATABASE AUDIT SPECIFICATION [Audit_Agent_Jobs]

FOR SERVER AUDIT [SIEM]

ADD (EXECUTE ON OBJECT::[dbo].[sp_add_job] BY [dbo])

WITH (STATE = ON)

 

-- DATABASE: Audit potentially dangerous procedures

-- Windows Log: Application

-- Events: 33205 

use master

CREATE DATABASE AUDIT SPECIFICATION [Audit_OSCMDEXEC]

FOR SERVER AUDIT [SIEM]

ADD (EXECUTE ON OBJECT::[dbo].[xp_cmdshell] BY [dbo]),                      -- Audit execution of "xp_cmdshell" 

ADD (EXECUTE ON OBJECT::[dbo].[sp_addextendedproc] BY [dbo]),               -- Audit execution of "custom extended stored procedures"

ADD (EXECUTE ON OBJECT::[dbo].[sp_execute_external_script] BY [dbo]),       -- Audit execution of external script such as  R et Python

ADD (EXECUTE ON OBJECT::[dbo].[Sp_oacreate] BY [dbo])                       -- Audit execution of "OLE Automation Procedure"

WITH (STATE = ON)
```

## Modules

For a metricset to go GA, the following criterias should be met:

* [ ] Supported versions are documented
* [ ] Supported operating systems are documented (if applicable)
* [ ] Integration tests exist
* [ ] System tests exist
* [ ] Automated checks that all fields are documented
* [ ] Documentation
* [ ] Fields follow [ECS](https://github.com/elastic/ecs) and [naming conventions](https://www.elastic.co/guide/en/beats/devguide/master/event-conventions.html)
* [ ] Dashboards exists (if applicable)
* [ ] Kibana Home Tutorial (if applicable)
  * Open PR against Kibana repo with tutorial. Examples can be found [here](https://github.com/elastic/kibana/tree/master/src/legacy/core_plugins/kibana/server/tutorials).

**Comment by peterdkdp** [comment]
Hi,

using this information, I created a workaround for this in our environment using a javascript processor.

**Procedure:**
1. Create a winlogbeat-sqlaudit.js file under C:\Program Files\winlogbeat\module\sqlaudit\config
2. Copy this code into the file:
```
// Purpose: parse the messages of MSSQL Audit events
// Created By: Peter De Keukelaere
// Created On: 20210216

var sqlaudit = (function () {
    var processor = require("processor");

    // MSSQL Audit Action Ids. This list was generated using select action_id, name from sys.dm_audit_actions
    //GROUP BY action_id, name
    //on a SQL 2019 server
    var auditActionIds = {
        "ACDO": "DATABASE_OBJECT_ACCESS_GROUP",
        "ACO ": "SCHEMA_OBJECT_ACCESS_GROUP",
        "ADBO": "BULK ADMIN",
        "ADDP": "DATABASE_ROLE_MEMBER_CHANGE_GROUP",
        "ADSC": "ADD SENSITIVITY CLASSIFICATION",
        "ADSP": "SERVER_ROLE_MEMBER_CHANGE_GROUP",
        "AL": "ALTER",
        "ALCN": "ALTER CONNECTION",
        "ALRS": "ALTER RESOURCES",
        "ALSS": "ALTER SERVER STATE",
        "ALST": "ALTER SETTINGS",
        "ALTR": "ALTER TRACE",
        "APRL": "ADD MEMBER",
        "AS": "ACCESS",
        "AUSC": "AUDIT SESSION CHANGED",
        "AUSF": "AUDIT SHUTDOWN ON FAILURE",
        "AUTH": "AUTHENTICATE",
        "BA": "BACKUP",
        "BAL ": "BACKUP LOG",
        "BCM ": "BATCH COMPLETED",
        "BCMG": "BATCH_COMPLETED_GROUP",
        "BRDB": "BACKUP_RESTORE_GROUP",
        "BST ": "BATCH STARTED",
        "BSTG": "BATCH_STARTED_GROUP",
        "C2OF": "TRACE AUDIT C2OFF",
        "C2ON": "TRACE AUDIT C2ON",
        "CCLG": "CHANGE LOGIN CREDENTIAL",
        "CMLG": "CREDENTIAL MAP TO LOGIN",
        "CNAU": "AUDIT_CHANGE_GROUP",
        "CO": "CONNECT",
        "CP": "CHECKPOINT",
        "CR": "CREATE",
        "D ": "DENY",
        "DABO": "DATABASE BULK ADMIN",
        "DAGF": "FAILED_DATABASE_AUTHENTICATION_GROUP",
        "DAGL": "DATABASE_LOGOUT_GROUP",
        "DAGS": "SUCCESSFUL_DATABASE_AUTHENTICATION_GROUP",
        "DBAF": "DATABASE AUTHENTICATION FAILED",
        "DBAS": "DATABASE AUTHENTICATION SUCCEEDED",
        "DBCC": "DBCC",
        "DBCG": "DBCC_GROUP",
        "DBL ": "DATABASE LOGOUT",
        "DL": "DELETE",
        "DPRL": "DROP MEMBER",
        "DR": "DROP",
        "DRSC": "DROP SENSITIVITY CLASSIFICATION",
        "DWC ": "DENY WITH CASCADE",
        "EX": "EXECUTE",
        "FT": "FULLTEXT",
        "FTG ": "FULLTEXT_GROUP",
        "G ": "GRANT",
        "GRDB": "DATABASE_PERMISSION_CHANGE_GROUP",
        "GRDO": "DATABASE_OBJECT_PERMISSION_CHANGE_GROUP",
        "GRO ": "SCHEMA_OBJECT_PERMISSION_CHANGE_GROUP",
        "GRSO": "SERVER_OBJECT_PERMISSION_CHANGE_GROUP",
        "GRSV": "SERVER_PERMISSION_CHANGE_GROUP",
        "GWG ": "GRANT WITH GRANT",
        "IMDP": "DATABASE_PRINCIPAL_IMPERSONATION_GROUP",
        "IMP ": "IMPERSONATE",
        "IMSP": "SERVER_PRINCIPAL_IMPERSONATION_GROUP",
        "IN": "INSERT",
        "LGB ": "BROKER LOGIN",
        "LGBG": "BROKER_LOGIN_GROUP",
        "LGDA": "DISABLE",
        "LGDB": "CHANGE DEFAULT DATABASE",
        "LGEA": "ENABLE",
        "LGFL": "FAILED_LOGIN_GROUP",
        "LGG ": "GLOBAL TRANSACTIONS LOGIN",
        "LGGG": "GLOBAL_TRANSACTIONS_LOGIN_GROUP",
        "LGIF": "LOGIN FAILED",
        "LGIS": "LOGIN SUCCEEDED",
        "LGLG": "CHANGE DEFAULT LANGUAGE",
        "LGM ": "DATABASE MIRRORING LOGIN",
        "LGMG": "DATABASE_MIRRORING_LOGIN_GROUP",
        "LGNM": "NAME CHANGE",
        "LGO ": "LOGOUT",
        "LGS ": "STORAGE LOGIN",
        "LGSD": "SUCCESSFUL_LOGIN_GROUP",
        "LGSG": "STORAGE_LOGIN_GROUP",
        "LO": "LOGOUT_GROUP",
        "MNDB": "DATABASE_CHANGE_GROUP",
        "MNDO": "DATABASE_OBJECT_CHANGE_GROUP",
        "MNDP": "DATABASE_PRINCIPAL_CHANGE_GROUP",
        "MNO ": "SCHEMA_OBJECT_CHANGE_GROUP",
        "MNSO": "SERVER_OBJECT_CHANGE_GROUP",
        "MNSP": "SERVER_PRINCIPAL_CHANGE_GROUP",
        "NMLG": "NO CREDENTIAL MAP TO LOGIN",
        "OP": "OPEN",
        "OPDB": "DATABASE_OPERATION_GROUP",
        "OPSV": "SERVER_OPERATION_GROUP",
        "PWAR": "APPLICATION_ROLE_CHANGE_PASSWORD_GROUP",
        "PWC ": "CHANGE PASSWORD",
        "PWCG": "LOGIN_CHANGE_PASSWORD_GROUP",
        "PWCS": "CHANGE OWN PASSWORD",
        "PWEX": "PASSWORD EXPIRATION",
        "PWMC": "MUST CHANGE PASSWORD",
        "PWPL": "PASSWORD POLICY",
        "PWR ": "RESET PASSWORD",
        "PWRS": "RESET OWN PASSWORD",
        "PWU ": "UNLOCK ACCOUNT",
        "R ": "REVOKE",
        "RC": "RECEIVE",
        "RCM ": "RPC COMPLETED",
        "RF": "REFERENCES",
        "RS": "RESTORE",
        "RST ": "RPC STARTED",
        "RWC ": "REVOKE WITH CASCADE",
        "RWG ": "REVOKE WITH GRANT",
        "SCCG": "SENSITIVITY_CLASSIFICATION_CHANGE_GROUP",
        "SL": "SELECT",
        "SN": "SEND",
        "SPLN": "SHOW PLAN",
        "STSV": "SERVER_STATE_CHANGE_GROUP",
        "SUQN": "SUBSCRIBE QUERY NOTIFICATION",
        "SVCN": "SERVER CONTINUE",
        "SVPD": "SERVER PAUSED",
        "SVSD": "SERVER SHUTDOWN",
        "SVSR": "SERVER STARTED",
        "TASA": "TRACE AUDIT START",
        "TASP": "TRACE AUDIT STOP",
        "TO": "TAKE OWNERSHIP",
        "TODB": "DATABASE_OWNERSHIP_CHANGE_GROUP",
        "TODO": "DATABASE_OBJECT_OWNERSHIP_CHANGE_GROUP",
        "TOO ": "SCHEMA_OBJECT_OWNERSHIP_CHANGE_GROUP",
        "TOSO": "SERVER_OBJECT_OWNERSHIP_CHANGE_GROUP",
        "TRBC": "TRANSACTION BEGIN COMPLETED",
        "TRBS": "TRANSACTION BEGIN STARTING",
        "TRCC": "TRANSACTION COMMIT COMPLETED",
        "TRCG": "TRACE_CHANGE_GROUP",
        "TRCS": "TRANSACTION COMMIT STARTING",
        "TRGC": "TRANSACTION PROPAGATE COMPLETED",
        "TRGS": "TRANSACTION PROPAGATE STARTING",
        "TRO ": "TRANSFER",
        "TRPC": "TRANSACTION PROMOTE COMPLETED",
        "TRPS": "TRANSACTION PROMOTE STARTING",
        "TRRC": "TRANSACTION ROLLBACK COMPLETED",
        "TRRS": "TRANSACTION ROLLBACK STARTING",
        "TRSC": "TRANSACTION SAVEPOINT COMPLETED",
        "TRSS": "TRANSACTION SAVEPOINT STARTING",
        "TX": "TRANSACTION_GROUP",
        "TXBG": "TRANSACTION BEGIN",
        "TXCG": "TRANSACTION_COMMIT_GROUP",
        "TXCM": "TRANSACTION COMMIT",
        "TXGG": "TRANSACTION_BEGIN_GROUP",
        "TXRB": "TRANSACTION ROLLBACK",
        "TXRG": "TRANSACTION_ROLLBACK_GROUP",
        "UCGP": "USER_CHANGE_PASSWORD_GROUP",
        "UDAG": "USER_DEFINED_AUDIT_GROUP",
        "UDAU": "USER DEFINED AUDIT",
        "UNDG": "STATEMENT_ROLLBACK_GROUP",
        "UNDO": "STATEMENT ROLLBACK",
        "UP": "UPDATE",
        "USAF": "CHANGE USERS LOGIN AUTO",
        "USLG": "CHANGE USERS LOGIN",
        "USTC": "COPY PASSWORD",
        "VDST": "VIEW DATABASE STATE",
        "VSST": "VIEW SERVER STATE",
        "VW": "VIEW",
        "VWCT": "VIEW CHANGETRACKING",
        "XA": "EXTERNAL ACCESS ASSEMBLY",
        "XU": "UNSAFE ASSEMBLY",
    };
    var classTypes = {
        "A": "SERVER AUDIT",
        "AF": "AGGREGATE",
        "AG": "AVAILABILITY GROUP",
        "AK": "ASYMMETRIC KEY",
        "AL": "ASYMMETRIC KEY LOGIN",
        "AP": "Undocumented",
        "AQ": "ADHOC QUERY",
        "AR": "APPLICATION ROLE",
        "AS": "ASSEMBLY",
        "AU": "ASYMMETRIC KEY USER",
        "BN": "REMOTE SERVICE BINDING",
        "C": "CHECK CONSTRAINT",
        "CD": "CREDENTIAL",
        "CK": "COLUMN ENCRYPTION KEY",
        "CL": "CERTIFICATE LOGIN",
        "CM": "COLUMN MASTER KEY",
        "CO": "SERVER CONFIG",
        "CP": "CRYPTOGRAPHIC PROVIDER",
        "CR": "CERTIFICATE",
        "CT": "CONTRACT",
        "CU": "CERTIFICATE USER",
        "D": "DEFAULT",
        "DA": "DATABASE AUDIT SPECIFICATION",
        "DB": "DATABASE",
        "DC": "DATABASE SCOPED CREDENTIAL",
        "DE": "DATABASE EVENT SESSION",
        "DK": "DATABASE ENCRYPTION KEY",
        "DN": "EVENT NOTIFICATION DATABASE",
        "DR": "DATABASE SCOPED RESOURCE GOVERNOR",
        "DS": "DATABASE SCOPED CONFIGURATION",
        "DT": "TRIGGER DATABASE",
        "DU": "AUDIT",
        "EC": "EDGE CONSTRAINT",
        "ED": "EXTERNAL DATA SOURCE",
        "EF": "EXTERNAL FILE FORMAT",
        "EL": "EXTERNAL LIBRARY",
        "EN": "EVENT NOTIFICATION",
        "EP": "ENDPOINT",
        "ER": "EXTERNAL RESOURCE POOL",
        "EX": "EXTERNAL SCRIPT QUERY",
        "F": "FOREIGN KEY CONSTRAINT",
        "FC": "FULLTEXT CATALOG",
        "FL": "FULLTEXT STOPLIST",
        "FN": "FUNCTION SCALAR SQL",
        "FP": "SEARCH PROPERTY LIST",
        "FS": "FUNCTION SCALAR ASSEMBLY ",
        "FT": "FUNCTION TABLE-VALUED ASSEMBLY ",
        "GU": "GROUP USER",
        "IF": "FUNCTION TABLE-VALUED INLINE SQL",
        "IS": "FUNCTION SCALAR INLINE SQL ",
        "IT": "INTERNAL TABLE",
        "IX": "INDEX",
        "LA": "EXTERNAL LANGUAGE",
        "LX": "LOGIN",
        "MK": "MASTER KEY",
        "MT": "MESSAGE TYPE",
        "OB": "OBJECT",
        "ON": "EVENT NOTIFICATION OBJECT",
        "P": "STORED PROCEDURE",
        "PC": "STORED PROCEDURE ASSEMBLY",
        "PF": "PARTITION FUNCTION",
        "PK": "PRIMARY KEY",
        "PL": "EXTERNAL GROUP LOGIN",
        "PQ": "PREPARED ADHOC QUERY",
        "PR": "BROKER PRIORITY",
        "PS": "PARTITION SCHEME",
        "PU": "EXTERNAL GROUP USER",
        "R": "RULE",
        "RF": "STORED PROCEDURE REPLICATION FILTER",
        "RG": "RESOURCE GOVERNOR",
        "RL": "ROLE",
        "RT": "ROUTE",
        "S": "TABLE SYSTEM",
        "SA": "SERVER AUDIT SPECIFICATION",
        "SC": "SCHEMA",
        "SD": "EVENT NOTIFICATION SERVER",
        "SE": "EVENT SESSION",
        "SG": "SERVER ROLE",
        "SK": "SYMMETRIC KEY",
        "SL": "SQL LOGIN",
        "SN": "SYNONYM",
        "SO": "SEQUENCE OBJECT",
        "SP": "SECURITY POLICY",
        "SQ": "QUEUE",
        "SR": "SERVER",
        "ST": "STATISTICS",
        "SU": "SQL USER",
        "SV": "SERVICE",
        "SX": "XML SCHEMA COLLECTION",
        "T": "TRIGGER SERVER",
        "TA": "TRIGGER ASSEMBLY",
        "TF": "FUNCTION TABLE-VALUED SQL",
        "TR": "TRIGGER",
        "TY": "TYPE",
        "U": "TABLE",
        "UQ": "UNIQUE CONSTRAINT",
        "US": "USER",
        "V": "VIEW",
        "WG": "WINDOWS GROUP",
        "WL": "WINDOWS LOGIN",
        "WU": "WINDOWS USER",
        "X": "STORED PROCEDURE EXTENDED",
        "XL": "EXTERNAL LOGIN",
        "XR": "XREL TREE",
        "XU": "EXTERNAL USER",
    };

    var parseAuditMessage = function(evt) {
        var msg = evt.Get("winlog.event_data.param1");
        if (!msg) {
            return;
        }
        var msgArray = msg.split("\n");
        var actionId;
        var sequenceNumber;
        var isColumnPermission;
        var sessionId;
        var serverPrincipalId;
        var databasePrincipalId;
        var targetServerPrincipalId;
        var targetDatabasePrincipalId;
        var objectId;
        var userDefinedEventId;
        var transactionId;
        var classType;
        var sequenceGroupId;
        var sessionServerPrincipalName;
        var serverPrincipalSid;
        var databasePrincipalName;
        var targetServerPrincipalName;
        var targetServerPrincipalSid;
        var targetDatabasePrincipalName;
        var serverInstanceName;
        var databaseName;
        var schemaName;
        var objectName;
        for (var i = 0; i < msgArray.length; i++)
        {
            var key = msgArray[i].split(":")[0];
            var value = msgArray[i].split(":")[1];
            switch (key)
            {
                case "sequence_number":
                    sequenceNumber = value;
                    break;
                case "action_id":
                    actionId = auditActionIds[value];
                    break;
                case "is_column_permission":
                    isColumnPermission = value;
                    break;
                case "session_id":
                    sessionId = value;
                    break;
                case "server_principal_id":
                    serverPrincipalId = value;
                    break;
                case "database_principal_id":
                    databasePrincipalId = value;
                    break;
                case "target_server_principal_id":
                    targetServerPrincipalId = value;
                    break;
                case "target_database_principal_id":
                    targetDatabasePrincipalId = value;
                    break;
                case "object_id":
                    objectId = value;
                    break;
                case "user_defined_event_id":
                    userDefinedEventId = value;
                    break;
                case "transaction_id":
                    transactionId = value;
                    break;
                case "class_type":
                    classType = classTypes[value];
                    break;
                case "sequence_group_id":
                    sequenceGroupId = value;
                    break;
                case "session_server_principal_name":
                    sessionServerPrincipalName = value;
                    break;
                case "server_principal_sid":
                    serverPrincipalSid = value;
                    break;                        
                case "database_principal_name":
                    databasePrincipalName = value;
                    break;                        
                case "target_server_principal_name":
                    targetServerPrincipalName = value;
                    break;                        
                case "target_server_principal_sid":
                    targetServerPrincipalSid = value;
                    break;                        
                case "target_database_principal_name":
                    targetDatabasePrincipalName = value;
                    break;                        
                case "server_instance_name":
                    serverInstanceName = value;
                    break;                        
                case "database_name":
                    databaseName = value;
                    break;                        
                case "schema_name":
                    schemaName = value;
                    break;
                case "object_name":
                    objectName = value;
                    break;
            }
        }

        evt.Put("event.action", actionId);
        evt.Put("event.sequence", sequenceNumber);
        evt.Put("mssql.audit.is_column_permission", isColumnPermission);
        evt.Put("mssql.audit.session_id", sessionId);
        evt.Put("mssql.audit.server_principal_id", serverPrincipalId);
        evt.Put("mssql.audit.database_principal_id", databasePrincipalId);
        evt.Put("mssql.audit.target_server_principal_id", targetServerPrincipalId);
        evt.Put("mssql.audit.target_database_principal_id", targetDatabasePrincipalId);
        evt.Put("mssql.audit.object_id", objectId);
        evt.Put("mssql.audit.user_defined_event_id", userDefinedEventId);
        evt.Put("mssql.audit.transaction_id", transactionId);
        evt.Put("mssql.audit.class_type", classType);
        evt.Put("mssql.audit.sequence_group_id", sequenceGroupId);
        evt.Put("mssql.audit.session_server_principal_name", sessionServerPrincipalName);
        evt.Put("mssql.audit.server_principal_sid", serverPrincipalSid);
        evt.Put("mssql.audit.database_principal_name", databasePrincipalName);
        evt.Put("mssql.audit.target_server_principal_name", targetServerPrincipalName);
        evt.Put("mssql.audit.target_server_principal_sid", targetServerPrincipalSid);
        evt.Put("mssql.audit.target_database_principal_name", targetDatabasePrincipalName);
        evt.Put("mssql.audit.server_instance_name", serverInstanceName);
        evt.Put("mssql.database.name", databaseName);
        evt.Put("mssql.database.schema_name", schemaName);
        evt.Put("mssql.database.object_name", objectName);
    };

    // Event ID 33205 - MSSQL Audit Event.
    var event33205 = new processor.Chain()
        .Add(parseAuditMessage)
        .Build();

    return {
        33205: event33205.Run,

        process: function (evt) {
            var event_id = evt.Get("winlog.event_id");
            var processor = this[event_id];
            if (processor === undefined) {
                throw "unexpected sqlaudit event_id";
            }
            evt.Put("event.module", "sqlaudit");
            processor(evt);
        },
    };
})();

function process(evt) {
    return sqlaudit.process(evt);
}
```
Add or remove fields in the declaration, switch statement and "evt.put" list if needed.

3. Update the winlogbeat.yml with if your audit logs are logged in the application log:
```
winlogbeat.event_logs:
  - name: Application
    ignore_older: 72h
    processors:
      - script:
          lang: javascript
          id: security
          file: ${path.home}/module/sqlaudit/config/winlogbeat-sqlaudit.js
          when.equals.winlog.event_id: 33205
  - name: Security
    ignore_older: 72h
    ...
```

But would be nice if this would be an official module, of course.

**Comment by nzjustin** [comment]
Thanks @peterdkdp, I have tested your processor script on Windows Server 2016, Microsoft SQL Server Enterprise, Filebeat 7.10.2. Works great.

+1 for a module

**Comment by drenze-athene** [comment]
+1 for module. I was looking for the same thing and was surprised that it didn't already exist.

**Comment by boomcubist** [comment]
@peterdkdp, I have followed your steps but I can't see the additional fields in Kibana. Are there any other steps that I need to follow?

**Comment by peterdkdp** [comment]
Did you check your index template? Maybe it's set to strict or not allowing dynamic fields.
Except for event.action and event.sequence the other fields are custom and are not available in ECS.
I've put them in an override template for winlogbeat in our environment.

**Comment by SigmazGFX** [comment]
Ahh found it.. 
The line in the winlogbeat.yml :
when.equals.winlog.event_id: 33205 

Had to be changed to singe quote the eventID
when.equals.winlog.event_id: '33205'

Stupid mistake on my part, I should have caught that.... All better now.
Thanks!

**Comment by SigmazGFX** [comment]
Added a conditional search to pull failed login usernames from the statement field. this makes it much easier to monitor the failed attempts when creating dashboards and lightens the load on Kibana by preprocessing the field in the .js instead of using a scripted field method.
`		        case "statement":
                    statement = value;
					  if(value.indexOf("Login failed for user") > -1) 
                    {
                        var start = value.indexOf("Login failed for user") + 23;
                        var end =  value.indexOf("'", start + 1);
                        evt.Put("mssql.audit.failed_user", value.substring(start, end));
					}else{
						statement = value;
					}
                    break;`

**Comment by peterdkdp** [comment]
> Added a conditional search to pull failed login usernames from the statement field. this makes it much easier to monitor the failed attempts when creating dashboards and lightens the load on Kibana by preprocessing the field in the .js instead of using a scripted field method.
> 
> > 
> 
> ```
>         case "statement":
>              statement = value;
> 	  if(value.indexOf("Login failed for user") > -1) 
>                 {
>                     var start = value.indexOf("Login failed for user") + 23;
>                     var end =  value.indexOf("'", start + 1);
>                     evt.Put("mssql.audit.failed_user", value.substring(start, end));
> 	}else{
>   	statement = value;
> 				}
>                 break;
> ```

I don't think this is really necessary, because from what I see that user name is already in **mssql.audit.server_principal_name**.
But it would be nice to have the mssql.database.name filled in when available.
For example:
- Login failed for user '_username_'. Reason: Password did not match that for the login provided. [CLIENT: _ipaddress_] **(no database)**
- Login failed for user '_username_'. Reason: Failed to open the explicitly specified database '_databasename_'. [CLIENT: _ipaddress_] **(database between quotes)**
- Login failed for user '_username_'. Reason: Could not find a user matching the name provided. [Database: '_databasename_'] [CLIENT: _ipaddress_] **(database between square brackets)**

**Comment by jamiehynds** [comment]
Just an FYI for folks following this MSSQL issue - we're currently working on an integration which supports MSSQL Audit events, via Elastic Agent - https://github.com/elastic/integrations/pull/2009

**Comment by jamiehynds** [comment]
Happy to announce that we've just shipped a new MSSQL integration with Elastic Agent to ingest SQL Audit Events (thanks to @r00tu53r). Thanks to everyone for chiming in on this issue and for providing various workarounds as we worked on the integration. If anyone is interested in testing out the integration and providing feedback, we'd love to hear it. 

I'll leave this issue open as we can discuss the integration here, if needed. 

Docs available here: https://github.com/elastic/integrations/tree/master/packages/microsoft_sqlserver/docs
